This patch fixes layout switching when 3 layouts are present.
It has been provided by Ivan Pascal in http://bugs.debian.org/345803

Not submitted upstream.

Index: xorg-server/xkb/xkbUtils.c
===================================================================
--- xorg-server.orig/xkb/xkbUtils.c
+++ xorg-server/xkb/xkbUtils.c
@@ -757,11 +757,11 @@
 
     grp= state->locked_group;
     if (grp>=ctrls->num_groups)
-	state->locked_group= XkbAdjustGroup(grp,ctrls);
+	state->locked_group= XkbAdjustGroup(XkbCharToInt(grp),ctrls);
 
     grp= state->locked_group+state->base_group+state->latched_group;
     if (grp>=ctrls->num_groups)
-	 state->group= XkbAdjustGroup(grp,ctrls);
+	 state->group= XkbAdjustGroup(XkbCharToInt(grp),ctrls);
     else state->group= grp;
     XkbComputeCompatState(xkbi);
     return;
